service/storage_service.cc

```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+    co_await _join_node_request_done.get_shared_future(_abort_source);
+
+    if (_join_node_response_done.available()) {
+        // We already handled this RPC. Return immediately for idempotence.
+        slogger.info("raft topology: the node got join_node_response RPC for the second time, ignoring");
+        co_return join_node_response_result{};
+    }
+
+    co_return co_await std::visit(overloaded_functor {
+        [&] (const join_node_response_params::accepted& acc) -> future<join_node_response_result> {
+            // Do a read barrier to read/initialize the topology state
+            auto& raft_server = _group0->group0_server();
+            co_await raft_server.read_barrier(&_abort_source);
+
+            // Calculate nodes to ignore
+            // TODO: ignore_dead_nodes setting for bootstrap
+            std::unordered_set<raft::server_id> ignored_ids;
+            auto my_request_it =
+                    _topology_state_machine._topology.req_param.find(_group0->load_my_id());
+            if (my_request_it != _topology_state_machine._topology.req_param.end()) {
+                if (auto* replace = std::get_if<service::replace_param>(&my_request_it->second)) {
+                    ignored_ids = replace->ignored_ids;
+                    ignored_ids.insert(replace->replaced_id);
+                }
+            }
+
+            // After this RPC finishes, repair or streaming will be run, and
+            // both of them require this node to see the normal nodes as UP.
+            // This condition might not be true yet as this information is
+            // propagated through gossip. In order to reduce the chance of
+            // repair/streaming failure, wait here until we see normal nodes
+            // as UP (or the timeout elapses).
+            std::vector<gms::inet_address> sync_nodes;
+            const auto& amap = _group0->address_map();
+            for (const auto& [id, _] : _topology_state_machine._topology.normal_nodes) {
+                if (ignored_ids.contains(id)) {
+                    continue;
+                }
+                if (auto addr = amap.find(id)) {
+                    sync_nodes.push_back(*addr);
+                }
+            }
+
+            slogger.info("raft topology: coordinator accepted request to join, "
+                    "waiting for nodes {} to be alive before responding and continuing",
+                    sync_nodes);
+            co_await _gossiper.wait_alive(sync_nodes, std::chrono::seconds(30));
+            slogger.info("raft topology: nodes {} are alive", sync_nodes);
+
+            // Unblock waiting raft_perform_join_handshake,
+            // which will start the raft server and continue
+            _join_node_response_done.set_value();
+
+            // Delay response until we are sure that this node will not retry
+            // the handshake on next boot.
+            co_await _join_node_group0_id_persisted.get_shared_future(_abort_source);
```

Comment:
Is this needed? What will happen without? The node will restart and fail to join because handshake will fail? But it is only if it restarts during bootstrap, no? And we decided the node will fail in such a case anyway,

Reply:
> Is this needed? What will happen without? The node will restart and fail to join because handshake will fail? But it is only if it restarts during bootstrap, no?

Yes, I agree that nothing bad should happen without it. Perhaps it's not worth complicating the control flow, so I can remove it.

Reply:
Actually, one bad thing could happen: the joining node, on restart, could receive a command to bootstrap itself before it sends `join_node_request` and receives an error. But that's most likely a different problem with a different fix, as the joining node should not accept such commands before a successful `join_node_request` (and we should start group0 server after that, too).

Reply:
Got rid of `_join_node_group0_id_persisted` promise, now server is started only after `join_node_request` finishes successfully, so the issue I described should no longer be relevant.

Reply:
> Got rid of `_join_node_group0_id_persisted` promise, now server is started only after `join_node_request` finishes successfully, so the issue I described should no longer be relevant.

But I see it is still here.

Reply:
But I do not see it if I checkout the pull request locally. This github thing is broken.

Reply:
Ah. This is because the line is removed by some later commit. But it means this commit cannot be compiled. Please check that each individual commit compiles.

Reply:
Done, also made sure that all commits compile in dev mode.